### DISCUSSION TITLE
Can `SdkError`s implement conversion to ServiceError or `aws_sdk_s3::Error` by reference and not value (move)?

### DISCUSSION THREAD

**Discussion by jrcavani** [awslabs/aws-sdk-rust#998] [General]
I'm implementing a simple flow of getting an object, and checking its errors - if it's a `NoSuchKey`, we log it one way and do something, and if it's a different variant, we log and do some other thing.

I'm having a hard time doing this beyond the suggested way (went through a few discussions/issues/Google searches):

```rust
    pub async fn get_object(&self, bucket: &str, key: &str) -> Result<Bytes> {
        let resp = self
            .client
            .get_object()
            .bucket(bucket)
            .key(key)
            .send()
            .await;
        let resp = match resp.map_err(|e| aws_sdk_s3::Error::from(e)) {
            Ok(resp) => Ok(resp),
            Err(aws_sdk_s3::Error::NoSuchKey(e)) => {
                warn!(
                    "Failed to get s3://{}/{}: {} (NoSuchKey). Skipping.",
                    bucket, key, e
                );
                Err(aws_sdk_s3::Error::NoSuchKey(e))
            }
            Err(e) => {
                warn!("Failed to get s3://{}/{}. Other error.", bucket, key);
                Err(e)
            }
        }?;

        let content = resp.body.collect().await?.into_bytes();
        anyhow::Ok(content)
    }
```

This works. However, sometimes I don't want to consume the `resp`. I would like to be able to take a reference of `resp: Result<GetObjectOutput, SdkError<...,...>>` and do error parsing without worrying about moving or reconstructing the `resp`.

```rust
    pub async fn get_object(&self, bucket: &str, key: &str) -> Result<Bytes> {
        let resp = self
            .client
            .get_object()
            .bucket(bucket)
            .key(key)
            .send()
            .await;
        match &resp.map_err(|e| aws_sdk_s3::Error::from(e)) {
            Err(aws_sdk_s3::Error::NoSuchKey(_)) => {
                warn!(
                    "Failed to get s3://{}/{}: (NoSuchKey). Skipping.",
                    bucket, key
                );
            }
            _ => {}
        };
        let resp = resp?;

        let content = resp.body.collect().await?.into_bytes();
        anyhow::Ok(content)
    }
```

The `SdkError` is not clone-able. It's `into_service_error()` also moves.

**Comment by jrcavani** [comment]
Oh I get it. So close...
```rust
    pub async fn get_object(&self, bucket: &str, key: &str) -> Result<Bytes> {
        let resp = self
            .client
            .get_object()
            .bucket(bucket)
            .key(key)
            .send()
            .await
            .map_err(aws_sdk_s3::Error::from);
        match &resp {
            Err(aws_sdk_s3::Error::NoSuchKey(_)) => {
                warn!(
                    "Failed to get s3://{}/{}: (NoSuchKey). Skipping.",
                    bucket, key
                );
            }
            _ => {}
        };
        let resp = resp?;

        let content = resp.body.collect().await?.into_bytes();
        anyhow::Ok(content)
    }
```

**Comment by jrcavani** [comment]
I'm here again... How do I catch 429 TooManyRequests?

After mapping to `aws_sdk_s3::Error`, it is variant `Unhandled`. To my surprise, the content, like `.meta()`, `.code()`, or `.message()` are all empty.

```rust
{:#} meta {:?} message {:?}
```

gives

```rust
unhandled error meta ErrorMetadata { code: None, message: None, extras: None }
```

**Reply by jrcavani** [reply]
How come aws-sdk-lambda has this https://docs.rs/aws-sdk-lambda/latest/aws_sdk_lambda/enum.Error.html#variant.TooManyRequestsException error variant, but S3 doesn't... Really need this to do infinite retries on some operations, and I'm currently getting by by catching all `Unhandled`.

**Reply by jdisanti** [reply]
In the worst case, you can fall back onto the `raw_response()` method on `SdkError`: https://docs.rs/aws-smithy-runtime-api/1.1.1/aws_smithy_runtime_api/client/result/enum.SdkError.html#method.raw_response

I don't know for sure, but my best guess is that S3 is responding with HTTP status 429 without a response body, so it doesn't turn into a modeled error, and that's why the meta/code/message fields are not populated.

**Comment by rcoh** [comment]
It seems like we need `.as_service_error() -> Option<&E>` to make it easier to check these things

**Reply by jrcavani** [reply]
Yeah I would like that. Otherwise I need to define my own errors to have more flexibility.

**Reply by rcoh** [reply]
this is done! Check out https://docs.rs/aws-smithy-runtime-api/latest/aws_smithy_runtime_api/client/result/enum.SdkError.html#method.as_service_error
